leetcodePython/SlidingWindow/slidingwindow_239.py

```diff
@@ -0,0 +1,27 @@
+from collections import deque
+from typing import List
+
+
+class Solution:
+    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
+        dq = deque()
+        res = []
+        for i in range(k):
+            # any data before i and smaller than nums[i] is not important anymore, nums[i] is only useful
+            while dq and nums[i] >= nums[dq[-1]]:
+                dq.pop()
+            dq.append(i)
+        # maximum for the 1st k elements
+        res.append(nums[dq[0]])
+
+        for i in range(k, len(nums)):
+            # out of k range
+            if dq and dq[0] == i-k:
+                dq.popleft()
+            # any data before i and smaller than nums[i] is not important anymore, nums[i] is only useful
+            while dq and nums[i] >= nums[dq[-1]]:
+                dq.pop()
```

Comment:
* Since the index iteration is from left to right.
* And the pop is happen every iteration
* This makes sure the maximum is at the left of the deque